ctms/acoustic_service.py

```diff
@@ -0,0 +1,279 @@
+        self.acoustic_main_table_id = acoustic_main_table_id
+        if isinstance(acoustic_newsletter_table_id, str):
+            acoustic_newsletter_table_id = int(acoustic_newsletter_table_id)
+        self.acoustic_newsletter_table_id = int(acoustic_newsletter_table_id)
+
+    def convert_ctms_to_acoustic(self, contact: ContactSchema):
+        acoustic_main_table = self._main_table_converter(contact)
+        newsletter_rows = self._newsletter_converter(acoustic_main_table, contact)
+        return acoustic_main_table, newsletter_rows
+
+    def _main_table_converter(self, contact):
+        acoustic_main_table = {
+            # populate with all the sub_flags set to false
+            # they'll get set to true below, as-needed
+            v: "0"
+            for v in AcousticResources.MAIN_TABLE_SUBSCR_FLAGS.values()
+        }
+        acceptable_subdicts = ["email", "amo", "fxa", "vpn_waitlist"]
+        special_cases = {
+            ("fxa", "fxa_id"): "fxa_id",
+            ("email", "primary_email"): "email",
+        }
+        for contact_attr in contact.dict():
+            subdict_value = getattr(contact, contact_attr)
+            contact_attr_name = str(contact_attr)
+            if contact_attr_name in acceptable_subdicts and subdict_value is not None:
+                for inner_attr in subdict_value.dict():
+                    inner_value = getattr(subdict_value, inner_attr)
+                    inner_attr_name = str(inner_attr)
+                    if contact_attr_name != "email":
+                        acoustic_field_name = contact_attr_name + "_" + inner_attr_name
+                    else:
+                        # Email table records should not be prefixed
+                        acoustic_field_name = inner_attr_name
+
+                    special_key = (contact_attr_name, inner_attr_name)
+                    if special_key in special_cases:
+                        acoustic_field_name = special_cases.get(special_key)
+
+                    if (
+                        acoustic_field_name
+                        in AcousticResources.VALID_ACOUSTIC_MAIN_TABLE_FIELDS
+                    ):
+                        acoustic_main_table[
+                            acoustic_field_name
+                        ] = self.transform_field_for_acoustic(inner_value)
+                    # else:
+                    # print(f"Skipping CTMS field ({contact_attr}, {inner_attr}) because no match "
+                    #       f"in Acoustic") # TODO: Convert to logging
+        return acoustic_main_table
+
+    def _newsletter_converter(self, acoustic_main_table, contact):
+        # create the RT rows for the newsletter table in acoustic
+        newsletter_rows = []
+        contact_newsletters: List[NewsletterSchema] = contact.newsletters
+        for newsletter in contact_newsletters:
+            if newsletter.name in AcousticResources.MAIN_TABLE_SUBSCR_FLAGS.keys():
```

Comment:
I wonder if we should log unhandled newsletters (ie- the "else" case here)?

Reply:
not urgent

Reply:
I wasn't sure how logging is handled with the background jobs... I can throw some basic logging in there on next update-commit.

Reply:
Added some basic logging, please re-review time-permitting. (resolving this, opening a new one would be easiest for me)